SiteTitle: Add bold and italic toolbar buttons for convenience

sejas opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

- Closes: #45854

## What?
<!-- In a few words, what is the PR actually doing? -->
Add Font face control to the SiteTitle toolbar where users can use `Bold` and `Italic` text styles:

<!-- old ![image](https://user-images.githubusercontent.com/36432/207459603-e8943cf6-1f6d-48a1-ab74-a2d939a9b120.png) -->

<img width="449" alt="Screenshot 2023-02-24 at 14 26 45" src="https://user-images.githubusercontent.com/779993/221202947-adbfaaaa-e1e3-4b61-9fa6-bfef091a6f98.png">


## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
The `SiteTitle` style options are difficult to access. I propose adding the bold and italic buttons to the toolbar allowing users to edit the text style easily.
In addition, we improve the accessibility of the block by allowing the shortcuts `Ctrl + B` to make the text bold and `Ctrl + I` to make it italic.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
I've added two new Toolbar buttons to the SiteTitle block. I'm using the `attribtues.style` as the source of truth of the bold and style state. For this reason, the new buttons and the Appearance Selector are in sync. Check the screencast to see it in action.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
1. Go to Appearance > Editor.
2. Click on `Edit`.
3. Select your `SiteTitle` block.
4. Observe that there are a new Typography button on the Toolbar, click it and observe the dropdown with the Bold and Italic options.
5. Click on the Bold button.
6. Observe the text transforms to bold.
7. Click on the Italic button.
8. Observe the text transforms to italic.
9. Observe that the Appearance select in the inner settings has changed too. The toolbar and the select are in sync.
10. Save the template.
11. Observe the Frontend of the site shows the correct styles.


### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
1. Select the `SiteTitle` block
8. Press `Ctrl + B` or `⌘ + B`
12. Observe the text gets bold.
13. Press `Ctrl + I` or `⌘ + I`
14. Observe the text gets italic.

## Screenshots or screencast <!-- if applicable -->

| **Before** | **After** |
|---|---|
| <img width="1178" alt="site-title-before" src="https://user-images.githubusercontent.com/779993/207203618-c669f32a-7566-4b17-8190-e9c221e327e9.png"> | <img width="1262" alt="Screenshot 2023-02-24 at 14 38 04" src="https://user-images.githubusercontent.com/779993/221205920-70a22e40-a829-484c-8024-c6f95c6cedcc.png"> |
<!-- old after <img width="1216" alt="site-title-after" src="https://user-images.githubusercontent.com/779993/207203652-68542345-2d93-41f2-8bed-88ddaf0c669d.png"> !-->

<!--
Old video using Bold and Italic icons.
https://user-images.githubusercontent.com/779993/207203733-3d1b898b-f989-4569-8cda-5f29f9a2b482.mp4
-->




https://user-images.githubusercontent.com/779993/220179387-4db637e5-9a68-44e6-b6fb-901e0b50c9a0.mp4
> The Typography icon was updated here https://github.com/WordPress/gutenberg/pull/46489#issuecomment-1443630583

sejas committed:
SiteTitle: add buttons to change the style appearance

sejas committed:
SiteTitle: handle default state for not bold and not italic

sejas committed:
SiteTitle: abstract onClick bold and italic

sejas committed:
SiteTitle: create Bold and ItalicControl components

sejas committed:
SiteTitle: add keyboard shortcuts and remove controls

danielbachhuber commented on packages/block-library/src/site-title/edit/use-typography-style.js:
```diff
@@ -0,0 +1,50 @@
+const BOLD = '700';
```
Do other blocks hard-code the bold value in this way?

sejas replied:
Yes. Here you can see an example.
- Bold: https://github.com/WordPress/gutenberg/blob/7796466b8a817477dd65c6fd6653c2259b62d828/packages/block-editor/src/components/font-appearance-control/index.js#L44-L47
- Italic: https://github.com/WordPress/gutenberg/blob/7796466b8a817477dd65c6fd6653c2259b62d828/packages/block-editor/src/components/font-appearance-control/index.js#L13-L16

Would you recommend doing something different?

The official CSS values can be found on MDN: [font-style](https://developer.mozilla.org/en-US/docs/Web/CSS/font-style#values) , [font-weight](https://developer.mozilla.org/en-US/docs/Web/CSS/font-weight#values)

danielbachhuber replied:
> Would you recommend doing something different?

Nope, I was just curious. Your approach seems fine 😊

aaronrobertshaw replied:
> Yes. Here you can see an example.

While the values do follow the CSS spec, given how tightly coupled this is to the block supports, it would be good to ensure these values can't get out of sync for any reason.

danielbachhuber reviewed (commented):
Awesome work, @sejas !

Overall, the buttons and keyboard shortcuts seem to work well.

I noticed a couple UX inconsistencies...

1. If the Site Title is already italicized, then it's not possible to remove the italicization:

<img width="1606" alt="image" src="https://user-images.githubusercontent.com/36432/207319743-198914bf-dfb9-400d-9ae9-8a41f78aeef8.png">

2. When you have some partially selected text, clicking "Bold" makes the entire text bold:

<img width="1597" alt="image" src="https://user-images.githubusercontent.com/36432/207320431-e3425ad5-3c49-4eea-ae17-87909c92b343.png">

3. Seen above, if you click "Bold", the inherited italicization is reverted.

The first UX inconsistency isn't new in this pull request, so I don't think we need to address it.

I personally think the second UX inconsistency is acceptable because the toolbar buttons are a UX enhancement for the majority of use cases. But, I wanted to call it out so it's stated.

The third UX inconsistency should probably be addressed.

sejas commented:
@danielbachhuber, thanks for testing it and adding your feedback.

> 1. If the Site Title is already italicized, then it's not possible to remove the italicization:

Did you add "Additional CSS" to your theme?
The user can add external CSS that modifies the appearance of any block. I think we shouldn't worry about it. If you add the same external CSS rule to a Paragraph, Can you undo the italicization using the toolbar buttons?.

> 2. When you have some partially selected text, clicking "Bold" makes the entire text bold:

Yes, that's right. The Bold and Italic buttons add inline CSS to the whole block changing the appearance/style of the entire text. We don't support partial bold or italic as RichText does because we would need to add HTML tags to the site title, changing the semantics and conflicting with the current Appearance selector that already exists.

> 3. Seen above, if you click "Bold", the inherited italicization is reverted.

I've tested it using the original SiteTitle, and I have experienced the same behavior. If I add external CSS to the theme, and I choose Bold using the "Appearance selector" the italicization is reverted to `regular` too.
<img width="147" alt="Screenshot 2022-12-13 at 17 18 48" src="https://user-images.githubusercontent.com/779993/207400505-67b47385-7ea0-4bda-adc9-b3760026963b.png">
The buttons behave identically as the "Appearance selector" does. This behavior already exists on SiteTitle, and this PR did not introduce it.
From my point of view, if the user/theme adds external CSS rules, then the behavior of the block can be compromised. Similar to point 1.
If you think it's necessary, I'm happy to revisit it or make a bigger refactor on the SiteTitle block.

danielbachhuber commented:
> Did you add "Additional CSS" to your theme? The user can add external CSS that modifies the appearance of any block. I think we shouldn't worry about it.

@sejas I didn't, no. I agree that we don't need to worry about it.

> The buttons behave identically as the "Appearance selector" does. This behavior already exists on SiteTitle, and this PR did not introduce it.
> From my point of view, if the user/theme adds external CSS rules, then the behavior of the block can be compromised. Similar to point 1.

Ah, ok. That's fine, then.

aaronrobertshaw reviewed (commented):
Excellent work on this @sejas 👍 

✅ Everything tests as advertised for me.

The generally low prominence of the font appearance control was by design. There’s some history in https://github.com/WordPress/gutenberg/issues/35604, but given the direction on the [original issue](https://github.com/WordPress/gutenberg/issues/45854#issuecomment-1341077638), I'd say the block toolbar additions and current behaviour are appropriate.

I don’t see any of the floated UX issues as blockers for this PR. They are potential opportunities to iterate further on the typography block supports in follow-ups.

> If the Site Title is already italicized, then it's not possible to remove the italicization:

The italic style should be able to be removed by selecting the "Regular" option in the font appearance control within the block inspector's typography panel. 

> Do other blocks hard-code the bold value in this way?

Given how tightly coupled a lot of this is to the typography block support, I wonder if it would be better to move some of this, such as the `useTypographyStyles` hook, to the typography block support and consume it from there. 

It might also avoid duplicating/syncing the bold value while allowing other blocks to potentially adopt the same feature as has been added to the Site Title block in this PR.

While exploring that option, perhaps it might be helpful to rename the hook to something more reflective of the subset of typography styles it is returning, e.g. `useFontAppearanceStyles`.

sejas commented:
@aaronrobertshaw, It's very smart to move the hook to the typography block support and share the bold/italic values to avoid future differences 👍 . Thanks for the suggestion 🙌.

I'm considering adding new keys to the `supports.typography` to show the new toolbar buttons. Other core blocks could use them just by enabling them. If we go this way, I will need to investigate how to add the shortcuts.

For example:

```diff
"supports": {
   ...
   "typography": {
	"fontSize": true,
	"lineHeight": true,
	"__experimentalFontFamily": true,
	"__experimentalTextTransform": true,
	"__experimentalTextDecoration": true,
	"__experimentalFontStyle": true,
	"__experimentalFontWeight": true,
	"__experimentalLetterSpacing": true,
	"__experimentalDefaultControls": {
		"fontSize": true,
		"lineHeight": true,
		"fontAppearance": true,
+		"fontAppearanceToolbar": true, // <- NEW. It would also check if `__experimentalFontStyle` or/and `__experimentalFontWeight` are enabled.
		"letterSpacing": true,
		"textTransform": true
	}
}

aaronrobertshaw commented:
> I'm considering adding new keys to the supports.typography to show the new toolbar buttons. Other core blocks could use them just by enabling them. If we go this way, I will need to investigate how to add the shortcuts.

That sounds like a logical next step to me 👍 

Might I suggest, though, that you move that exploration to a follow-up PR? 

That way, the great work already done in this PR can land sooner and won't get bogged down by discussions around how best to fit the proposed change into the shape of the block supports, what blocks could or should benefit from it etc.

sejas committed:
SiteTitle: export useFontAppearanceStyles from block-editor hooks

sejas committed:
SiteTitle: move useFontAppearanceStyles to a separate file

sejas committed:
SiteTitle: import __experimentalUseFontAppearanceStyles hook from @wordpress/block-editor

sejas committed:
SiteTitle: move shortcuts logic into the hook onKeyDownToggleFontAppearance

sejas committed:
SiteTitle: remove test code

sejas committed:
SiteTitle: remove useTypographyStyle in favor of useFontAppearanceStyles

sejas committed:
SiteTitle: declare onKeyDownToggleFontAppearance as a function

sejas commented:
> Might I suggest, though, that you move that exploration to a follow-up PR?

I agree. Let's finish this PR first 👌.

I've updated this PR with the following changes:

- Renamed the hook to `useFontAppearanceStyles` and moved to `block-editor/src/hooks`. I've decided to create a separate file. Another good place could be inside `font-appearance.js`.
- I'm consuming the hook with the prefix `__experimental`.
- `font-appearance-control` and the new hook use the same variables for Bold and Italic.
- Moved the keydown/shotrcut logic inside the hook.

I still need to change the icons to differentiate them from the RichText buttons. I'll post some design suggestions  in the issue: https://github.com/WordPress/gutenberg/issues/45854#issuecomment-1352034503

aaronrobertshaw commented on packages/block-library/src/site-title/edit/index.js:
```diff
@@ -15,10 +15,13 @@ import {
 	InspectorControls,
 	BlockControls,
 	useBlockProps,
+	__experimentalUseFontAppearanceStyles,
```
```suggestion
	__experimentalUseFontAppearanceStyles as useFontAppearanceStyles,
```

I think the pattern usually adopted in Gutenberg is to import the experimental API under what its stabilized name would be to reduce future code changes and possibly prevent some missed updates.

sejas replied:
Done!, thanks for the suggestion.

aaronrobertshaw commented on packages/block-library/src/site-title/edit/index.js:
```diff
@@ -51,12 +54,21 @@ export default function SiteTitleEdit( {
+		toggleBold,
+		toggleItalic,
+		onKeyDownToggleFontAppearance,
+	} = __experimentalUseFontAppearanceStyles( { attributes, setAttributes } );
```
```suggestion
	} = useFontAppearanceStyles( { attributes, setAttributes } );
```
If adopting the pattern suggested in my comment on the `__experimentalUseFontAppearanceStyles` import, this would need a tweak as well.

sejas replied:
Done! https://github.com/WordPress/gutenberg/pull/46489/commits/2a7ca22149c7a2b1d095b9b294d66c30c66172b6

aaronrobertshaw commented on packages/block-editor/src/hooks/use-font-appearance-styles.js:
```diff
@@ -0,0 +1,81 @@
+	const isBold = typographyStyle?.fontWeight === FONT_WEIGHT_BOLD;
+	const isItalic = typographyStyle?.fontStyle === FONT_STYLE_ITALIC;
+
+	function getNewTypographyStyle( newIsBold, newIsItalic ) {
```
Would renaming this to something more closely related to the hook's name and its purpose around font appearance improve clarity?

sejas replied:
yes!, it makes total sense.
Done: https://github.com/WordPress/gutenberg/pull/46489/commits/3c03f4b1f4364b6fa8b8a9c4a3bc9093b90fd158

aaronrobertshaw commented on packages/block-library/src/site-title/edit/index.js:
```diff
@@ -15,10 +15,13 @@ import {
+import { ToggleControl, PanelBody, ToolbarButton } from '@wordpress/components';
 import { createBlock, getDefaultBlockName } from '@wordpress/blocks';
 import { decodeEntities } from '@wordpress/html-entities';
+import { formatBold, formatItalic } from '@wordpress/icons';
```
As noted earlier, some icons distinct from the rich text bold and italic ones would be good here.

aaronrobertshaw reviewed (commented):
Thank you for the continued iterations on this PR @sejas 👍 

It is still testing well for me.

The one thing I'm not certain about is the introduction of the `__experimental` API. I'm aware that there's a concerted effort to avoid introducing experimental APIs as they previously had a habit of working their way into core.

Recently, a new approach to managing experiments was merged via https://github.com/WordPress/gutenberg/pull/43386. I get the feeling that might be overkill for this particular case, though.

Perhaps we could get additional reviews and input on this PR to avoid the experimental stage for the new custom hook altogether. This might also be able to occur while we're waiting on the new bold and italic icons for the toolbar buttons.

What do you think?

sejas committed:
SiteTitle: import hook without experimental prefix

sejas committed:
SiteTitle: rename font appearance function

sejas commented:
> Perhaps we could get additional reviews and input on this PR to avoid the experimental stage for the new custom hook altogether. This might also be able to occur while we're waiting on the new bold and italic icons for the toolbar buttons.
>
> What do you think?

Yes, I think it would be a great idea to avoid the experimental stage for the new hook 👌

kozer commented:
@sejas very nice work!!!

Everything looks nice. Really loved the constants that you used for bold, italic etc..

As @danielbachhuber , I have the same UI inconsistencies:
1. If the Site Title is already italicized, then it's not possible to remove the italicization
2. When you have some partially selected text, clicking "Bold" makes the entire text bold (This is not happening in the Custom link block for example)
3. If you click "Bold", the inherited italicization is reverted.

I also noticed that by changing the style back to "Regular", the italics are being removed. and pressing the button again, we end up in the 1 case.

Again, nice work!

sejas committed:
SiteTitle: refactor font appearance inside a toolbar dropdown menu

sejas committed:
SiteTitle: use isActive instead of isSelected control attribute

sejas committed:
SiteTitle: add isPressed icon status to font appearance control

sejas committed:
SiteTitle: move FontAppearanceControl to other toolbar group

sejas committed:
SiteTitle: fix typo on appearance

jameskoster commented:
Thanks for the update @sejas, are you using the `typography` icon from the [WordPress Icons package](https://wordpress.github.io/gutenberg/?path=/story/icons-icon--library)? It looks a bit different.

sejas committed:
SiteTitle: update typography icon

sejas commented:
> Thanks for the update @sejas, are you using the `typography` icon from the [WordPress Icons package](https://wordpress.github.io/gutenberg/?path=/story/icons-icon--library)? It looks a bit different.

Oh!, I used `formatUppercase` icon.
Fixed here https://github.com/WordPress/gutenberg/pull/46489/commits/fa2b28bf9a834d51df4fb52d710417cf73261262 👌. 
@jameskoster, Thanks for catching the mistake.

<img width="841" alt="Screenshot 2023-02-24 at 12 37 24" src="https://user-images.githubusercontent.com/779993/221181925-a2072e45-964e-4de6-b4ac-9f3deb8a1b19.png">

jameskoster commented:
Thanks for changing that detail :)  

The icon isn't the perfect one, but I'm struggling to think of a better one right now. For me it's probably okay to continue moving forwards with this one, and circle back to explore a new icon later. But I'd appreciate another set of eyes from @WordPress/gutenberg-design in case I'm missing an obvious option.

jasmussen commented:
Happy to try this and revisit.

jameskoster commented:
I just took this for a spin:

<img width="461" alt="Screenshot 2023-02-27 at 11 54 29" src="https://user-images.githubusercontent.com/846565/221557324-d666f561-7205-44c4-ae3a-94b736811635.png">

Bold doesn't appear to be working, could that be something on my local setup?

I'm not sure we need to the 'toggled' state on the toolbar button when styles are selected.

sejas commented:
@jameskoster

> Bold doesn't appear to be working, could that be something on my local setup?

Does the Bold style work for you when using the original "Typography panel" ?

<img width="250" alt="Screenshot 2023-02-27 at 12 30 41" src="https://user-images.githubusercontent.com/779993/221564432-25ca88ae-9aaf-4b5c-9dfb-7056930ee4d6.png">


> I'm not sure we need to the 'toggled' state on the toolbar button when styles are selected.

Happy to remove it if needed. I intended to make clear that some state is applied inside. The heading and paragraph buttons use the Value Icon to represent this state. But we can't use the `B` _`I`_ icons, so I went for the toggled versions with black background.

jameskoster commented:
> Does the Bold style work for you when using the original "Typography panel" ?

No. I guess that's an issue on my local :)

> Happy to remove it if needed

Yeah I don't think it's really necessary.

danielbachhuber committed:
Merge branch 'trunk' into add/site-title-bold-italic-styles

danielbachhuber reviewed (approved):
Works well to my testing!

https://user-images.githubusercontent.com/36432/221698406-76fc080c-a0ee-4c15-8f62-ad5b86c0ce14.mp4

FYI - I merged `trunk`.

Probably outside the scope of this PR, but it might be cool if the icon reflected the font appearance (e.g. appeared as bolded when "Bold" was selected, italicized when "Italic" was selected, etc.)

I'll defer to the @WordPress/gutenberg-core folks for final approval on the code.

youknowriad commented on packages/block-library/src/site-title/edit/index.js:
```diff
@@ -15,6 +15,7 @@ import {
 	InspectorControls,
 	BlockControls,
 	useBlockProps,
+	__experimentalUseFontAppearanceStyles as useFontAppearanceStyles,
```
Any reason for this to leave in the block editor if it's only used in the site-title block?

Maybe folks working on block tools can shed more light on this new toolbar. cc @aaronrobertshaw @andrewserong @jasmussen

youknowriad replied:
Ok, just saw that @aaronrobertshaw is already reviewing this PR :)

aaronrobertshaw replied:
👍 Re-reviewing this PR is on my radar for the next couple of days.

The logic behind the hook living in the block-editor package was that it is so closely tied to the typography block supports in that package. There's also the potential for it to be re-used across other blocks leveraging the font weight and style supports.

youknowriad replied:
Maybe we can start with it being "private" unless we're certain that it's an API we want to offer to third-party devs.

aaronrobertshaw reviewed (commented):
Thanks for continuing to iterate on this one @sejas 👍 

In its current state, the PR is testing very nicely for me. Great work!

Before landing this, there are a couple more things we should probably address.

1. [Removing the toggled state](https://github.com/WordPress/gutenberg/pull/46489#issuecomment-1446712409) for the block toolbar's typography dropdown.
2. Now we have the capability to `lock` and `unlock` experimental APIs, I think the best and quickest path to landing this will be to leverage that as @youknowriad [suggested](https://github.com/WordPress/gutenberg/pull/46489/files#r1119720145). You can find more info on this in the [coding guidelines](https://github.com/WordPress/gutenberg/blob/7fbf73d447b211c31d75585c46fb3f08e6267582/docs/contributors/code/coding-guidelines.md#use-the-lock-and-unlock-api-from-wordpressprivate-apis-to-privately-export-almost-anything).
3. _(Optional)_ If the intent is for more than just the Site Title block to eventually offer the block toolbar control for font appearance, moving the new `FontAppearanceControl` to a component alongside the hook so it can be reused might make sense. If doing this, it should be locked as private/experimental as well. 

With the above tweaks addressed, I think this PR will be in good shape. ✨
